src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubectlBuildWrapper.java

```diff
@@ -119,58 +93,43 @@ public void setUp(Context context, Run<?, ?> build, FilePath workspace, Launcher
         }
 
-        if (c instanceof FileCredentials) {
-            try (InputStream in = ((FileCredentials) c).getContent(); OutputStream out = configFile.write()) {
-                IOUtils.copy(in, out);
+        // create Kubeconfig
+        if (auth instanceof KubernetesAuthKubeconfig) {
+            try (Writer w = new OutputStreamWriter(new FileOutputStream(configFile.getRemote()), "UTF-8")) {
+                w.write(((KubernetesAuthKubeconfig) auth).getKubeconfig());
             }
             return;
         }
 
-        if (c instanceof StringCredentials) {
-            login = "--token=" + ((StringCredentials) c).getSecret().getPlainText();
-        } else if (c instanceof TokenProducer) {
-            login = "--token=" + ((TokenProducer) c).getToken(serverUrl, null, true);
-        } else if (c instanceof UsernamePasswordCredentials) {
-            UsernamePasswordCredentials upc = (UsernamePasswordCredentials) c;
-            login = "--username=" + upc.getUsername() + " --password=" + Secret.toString(upc.getPassword());
-        } else if (c instanceof StandardCertificateCredentials) {
-            StandardCertificateCredentials scc = (StandardCertificateCredentials) c;
-            KeyStore keyStore = scc.getKeyStore();
-            String alias;
-            try {
-                alias = keyStore.aliases().nextElement();
-                X509Certificate certificate = (X509Certificate) keyStore.getCertificate(alias);
-                Key key = keyStore.getKey(alias, Secret.toString(scc.getPassword()).toCharArray());
-                FilePath clientCrtFile = workspace.createTempFile("client", "crt");
-                FilePath clientKeyFile = workspace.createTempFile("client", "key");
-                String encodedClientCrt = wrapWithMarker(BEGIN_CERTIFICATE, END_CERTIFICATE,
-                        Base64.encodeBase64String(certificate.getEncoded()));
-                String encodedClientKey = wrapWithMarker(BEGIN_PRIVATE_KEY, END_PRIVATE_KEY,
-                        Base64.encodeBase64String(key.getEncoded()));
-                clientCrtFile.write(encodedClientCrt, null);
-                clientKeyFile.write(encodedClientKey, null);
-                tempFiles.add(clientCrtFile.getRemote());
-                tempFiles.add(clientKeyFile.getRemote());
-                login = "--client-certificate=" + clientCrtFile.getRemote() + " --client-key="
-                        + clientKeyFile.getRemote();
-            } catch (KeyStoreException e) {
-                throw new AbortException(e.getMessage());
-            } catch (UnrecoverableKeyException e) {
-                throw new AbortException(e.getMessage());
-            } catch (NoSuchAlgorithmException e) {
-                throw new AbortException(e.getMessage());
-            } catch (CertificateEncodingException e) {
-                throw new AbortException(e.getMessage());
-            }
+        String login;
+        if (auth instanceof KubernetesAuthToken) {
```

Comment:
Would it make sense to replace all this code with polymorphism and a method to implement in `KubernetesAuth` `public String generateKubectlConfigArguments()`

Reply:
Sounds good, but not sure, how to implement it for KubernetesAuthCertificate - it requires some file paths, which are generated in KubectlBuildWrapper

Reply:
The most naive option would be to add `FilePath workspace` to the method signature. Could pass the whole `Context` object and use `Context.get(FilePath.class)` to retrieve the workspace. This would however bind `KubernetesAuth` to Pipeline usage which is not ideal.

Reply:
It would take much effort to rework that part, there are few problems. These arguments required only once to create kubeconfig file, and since KubernetesAuthKubeconfig is one of the possible authentication methods, this case should be handled separately. So, I am not sure, what should be returned from `KAuthKubeconfig.generateKubectlConfigArguments` ?

Both Kubeconfig and Certificate auths rely on temp files, that should be cleaned after build. So we would need to pass `Set<String> tempFiles` together with `FilePath workspace`, which doesn't also seem to be clearest solution.